### DISCUSSION TITLE
Setup for storing large volume of durable sessions

### DISCUSSION THREAD

**Discussion by RomekCDPR** [keycloak/keycloak#25170] [Q&A]
Hello,

My team is planning to implement solution based on the Keycloak. I'd like to ask the community for a feedback regarding possible setups.

# Requirements

1. Long session lifespan (1 year)
2. Session durability - surviving nodes restarts.
3. Estimated numer of sessions: 100-500M.
4. Common custom providers updates - ideally we would like to perform zero-downtime rolling updates (we are aware that we would need to have service downtime in case of Keycloak upgrade - https://github.com/keycloak/keycloak/discussions/21067#discussioncomment-6254291)

# Alternatives

Diagrams represent only 1 node for readability.

## Embedded Infinispan with [JDBC persistence](https://infinispan.org/docs/stable/titles/configuring/configuring.html#jdbc-connection-factories_persistence)

![embedded_jdbc](https://github.com/keycloak/keycloak/assets/140800159/3b912586-971a-44e8-9bc7-57c730ae9db4)

Pros:
- No extra Infinispan cluster to maintain

Cons:
- High memory requirements (estimated 0.5 Ti for 100M sessions).
- Updating nodes requires long rebalancing that can increase deployment/scaling time and service availability.
- Keycloak skips writing to DB for some of the operations. [Keycloak 23 should fix it](https://github.com/keycloak/keycloak/commit/a45934a762d8726d1cd3ea9c2c0e34e09dd5b849).
- Risk of persisted session object compatibility after Keycloak upgrade.
- Additional DB/tables to maintain.

## Embedded Infinispan with [eviction](https://infinispan.org/docs/stable/titles/configuring/configuring.html#eviction_configuring-memory-usage) and [JDBC persistence](https://infinispan.org/docs/stable/titles/configuring/configuring.html#jdbc-connection-factories_persistence)

![embedded_eviction_jdbc](https://github.com/keycloak/keycloak/assets/140800159/6b489743-3f02-4d57-bbdf-b2f1f1424147)

Pros:
- No extra Infinispan cluster to maintain
- Low memory requirements.
- Quick nodes updates.

Cons:
- Higher latency (DB read/writes)
- Keycloak skips writing to DB for some of the operations. [Keycloak 23 should fix it](https://github.com/keycloak/keycloak/commit/a45934a762d8726d1cd3ea9c2c0e34e09dd5b849).
- Keycloak skips loading from DB for some of the operations (e.g. listing user sessions).
- Risk of persisted session object compatibility after Keycloak upgrade.
- Additional DB/tables to maintain.
- Eviction/expiration in embedded cache leads to `Failed to replace entity warnings` + high latency. 

## Embedded Infinispan + offline sessions

![embedded_offline](https://github.com/keycloak/keycloak/assets/140800159/31b39740-94eb-4648-92bf-e9391e7ece03)

Pros:
- "Default" setup.
- No extra Infinispan cluster to maintain.
- No extra DB to maintain.

Cons:
- High memory requirements (estimated 1 Ti for 100M sessions).
- Higher latency (DB read/writes).
- Updating nodes requires long rebalancing that can increase deployment/scaling time and service availability.
- Abusing offline sessions side effects.

## Embedded Infinispan with [eviction](https://infinispan.org/docs/stable/titles/configuring/configuring.html#eviction_configuring-memory-usage)/[expiration](https://infinispan.org/docs/stable/titles/configuring/configuring.html#expiration_configuring-memory-usage) + offline sessions

![embedded_eviction_offline](https://github.com/keycloak/keycloak/assets/140800159/d33b61e0-d3be-41c6-802a-a9ba8762654c)

Can be more feasible with https://github.com/keycloak/keycloak/discussions/24274.

Pros:
- No extra Infinispan cluster to maintain.
- No extra DB to maintain.
- Low memory requirements.
- Quick nodes updates.

Cons:
- Higher latency (DB read/writes).
- Abusing offline sessions side effects.
- Eviction from heap memory uses LFU algorithm that evicts freshly loaded sessions.
- Eviction/expiration in embedded cache leads to `Failed to replace entity warnings` + high latency. 

## Embedded Infinispan + [cross-site replication](https://infinispan.org/docs/stable/titles/xsite/xsite.html)

![embedded_xsite](https://github.com/keycloak/keycloak/assets/140800159/093abc24-d136-4459-a6c6-0e4336a19c8a)

Cons:
- High memory requirements (estimated 0.5 Ti for 100M sessions).
- Additional Infinispan cluster maintenance.
   - It should be only 1 small instance.
- Updating nodes requires long rebalancing that can increase deployment/scaling time and service availability.
- Risk of persisted session object compatibility after Keycloak upgrade.
- Additional DB to maintain.
- Higher latency (DB read/writes).

## Embedded Infinispan with [eviction](https://infinispan.org/docs/stable/titles/configuring/configuring.html#eviction_configuring-memory-usage)/[expiration](https://infinispan.org/docs/stable/titles/configuring/configuring.html#expiration_configuring-memory-usage) + remote Infinispan with [JDBC persistence](https://infinispan.org/docs/stable/titles/configuring/configuring.html#jdbc-connection-factories_persistence)

![embeded_eviction_remote_jdbc](https://github.com/keycloak/keycloak/assets/140800159/f1493128-7295-4713-8da8-ef762e3c2b1d)

Pros:
- Low memory requirements.
- Quick nodes updates.

Cons:
- Additional Infinispan cluster maintenance
- Risk of persisted session object compatibility after Keycloak upgrade.
- Additional DB to maintain.
- Higher latency (DB read/writes).
- Keycloak skips loading from remote cache for some of the operations (e.g. listing user sessions).
- Eviction/expiration in embedded cache leads to `Failed to replace entity warnings` + high latency. 

## Embedded Infinispan with [eviction](https://infinispan.org/docs/stable/titles/configuring/configuring.html#eviction_configuring-memory-usage) + [remote Infinispan](https://www.keycloak.org/high-availability/connect-keycloak-to-external-infinispan) with [eviction](https://infinispan.org/docs/stable/titles/configuring/configuring.html#eviction_configuring-memory-usage)/[expiration](https://infinispan.org/docs/stable/titles/configuring/configuring.html#expiration_configuring-memory-usage) + offline sessions

![embedded_eviction_remote_eviction_offline](https://github.com/keycloak/keycloak/assets/140800159/a29efa95-ffe5-457b-b744-09bab602d8a8)

Pros:
- Low memory requirements.
- Quick nodes updates.

Cons:
- Higher latency (DB read/writes) 
    - Should be smaller than "Embedded Infinispan with eviction/expiration + offline sessions" (we can keep more sessions in the memory and have quick nodes updates).
- Abusing offline sessions side effects.
- Eviction from heap memory uses LFU algorithm that evicts freshly loaded sessions.
- Eviction/expiration in embedded cache leads to `Failed to replace entity warnings` + high latency. 

## Embedded Infinispan with [eviction](https://infinispan.org/docs/stable/titles/configuring/configuring.html#eviction_configuring-memory-usage) + [remote Infinispan](https://www.keycloak.org/high-availability/connect-keycloak-to-external-infinispan) + [cross-site replication](https://infinispan.org/docs/stable/titles/xsite/xsite.html)

![embedded_eviction_remote_xsite](https://github.com/keycloak/keycloak/assets/140800159/3f211614-5f7a-40ec-9d50-62ea18630d31)

Pros:
- Quick nodes updates.

Cons:
- High memory requirements (estimated 0.5 Ti for 100M sessions).
- 2 additional Infinispan cluster maintenance.
   - 1 of them should be only 1 small instance.
- Risk of persisted session object compatibility after Keycloak upgrade.
- Additional DB to maintain.
- Keycloak skips loading from remote cache for some of the operations (e.g. listing user sessions).
- Higher latency (DB read/writes) .
- Eviction/expiration in embedded cache leads to `Failed to replace entity warnings` + high latency. 

## Custom `UserSessionProvider`

Custom DB persistence with optional in-memory cache optimisations.

Pros:
- Full control of the logic.

Cons:
- Requires implementation and maintenance of `UserSessionProvider` and `UserSessionModel`.
- High cost of upgrading Keycloak (requires large code changes).
- Higher latency (DB read/writes).

# Outcome

We are leaning towards "Custom `UserSessionProvider`" - the only alternative that meets all of the requirements - even with hight implementation/maintenance costs.

**Comment by sirishkumar** [comment]
@RomekCDPR Regarding your comment "Eviction/expiration in embedded cache leads to Failed to replace entity warnings + high latency." .  There is a discussion may be related to this at https://github.com/keycloak/keycloak/discussions/16405 . Came across your question while I am trying to find the root cause. It would be great if you can provide some details why using embedded cache leads to "Failed to replace entity" warning.

**Reply by RomekCDPR** [reply]
Hello @sirishkumar - we observed the problem in setups with embedded cache storing subset of all user sessions - e.g.:

1. Embedded Infinispan with eviction + JPA persistence
2. Embedded Infinispan with eviction/expiration + offline sessions
3. Embedded Infinispan with eviction/expiration + remote Infinispan

My hypothesises would be that Keycloak assumes that all users sessions are in embedded cache - it tries to update it in a loop: https://github.com/keycloak/keycloak/blob/ee126f73e532baaabd8c82b94df8c70031ff650b/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/InfinispanChangelogBasedTransaction.java#L225
which causes the hight latency and emits warnings.

**Reply by sirishkumar** [reply]
We store all sessions only in the embedded infinispan and we still have this warnings.

**Comment by ahus1** [comment]
Since this question was asked, several changes happend to Keycloak. 

The setup asked for in the question is now supported since Keycloak 26.0 which was released in October 2024: Since that release Keycloak by default stores all sessions in the database. 

You are no longer limited by memory usage. Depending on the logins per second and token refreshes per second, you might be limited by DB IO.

**Reply by RomekCDPR** [reply]
We ended up implementing custom `UserSessionProvider` last year. It performs well for the traffic we are receiving. We are aware of the changes in 26.0, and plan to review the `PersistentUserSessionProvider` and run stress/soak tests on it. If it performs well we would consider migrating from our custom sulution.
